app/scripts/modules/core/pipeline/config/pipelineConfigProvider.js

```diff
@@ -59,10 +59,45 @@ module.exports = angular.module('spinnaker.core.pipeline.config.configProvider',
       return angular.copy(stageTypes);
     }
 
-    function getConfigurableStageTypes() {
-      return getStageTypes().filter(function(stageType) {
-        return !stageType.synthetic && !stageType.provides;
+    function getConfigurableStageTypes(providers) {
+      let allStageTypes = getStageTypes();
+      let configurableStageTypes = allStageTypes.filter(stageType => !stageType.synthetic && !stageType.provides);
+      if (!providers) {
+        return configurableStageTypes;
+      }
+      addCloudProviders(allStageTypes, configurableStageTypes, providers);
+      return configurableStageTypes
+        .filter(stageType => stageType.cloudProviders.length)
+        .sort((a, b) => a.label.localeCompare(b.label));
+    }
+
+    function addCloudProviders(allStageTypes, configurableStageTypes, providers) {
+      configurableStageTypes.forEach(type => {
```

Comment:
Might be a bigger discussion needed, but I would like to see a more "functional" approach here.  Instead of looping over a collection and mutating the underlying data structures we should be mapping over the existing list creating a new augmented list.